### PR TITLE
Expand what crm_rule can check

### PR DESCRIPTION
**Pull request by clumens**
I'm not really convinced the CRM_EX_UNSATISFIED patch is necessary, but it seemed worth it to differentiate between the expired/unexpired and currently-in-effect/not-in-effect cases.  They seem subtly different to me.

### CODE REVIEW

File: tools/crm_rule.c [hunk 1]
```diff
@@ -63,16 +63,19 @@ crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effe
...
-    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation!='date_spec']", rule_id);
+    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation!='date_spec'] |"
+                              "//rule[@id='%s']/date_expression[@operation='date_spec' and date_spec/@years and not(date_spec/@moon)]",
+                              rule_id, rule_id);
     xpathObj = xpath_search(cib_constraints, xpath);
 
     max = numXpathResults(xpathObj);
     if (max == 0) {
-        CMD_ERR("No rule found with ID=%s containing a date_expression", rule_id);
+        CMD_ERR("No rule found with ID=%s, or rule does not have a date_expression, or rule is too complicated", rule_id);
```
Comment (kgaillot) [inline tools/crm_rule.c]: I'm thinking a more specific message would be worth the overhead of more xpath searches, which shouldn't really be noticeable since crm_rule will only ever be called manually (vs in a loop of some sort or at a time-sensitive moment). We could even potentially do the extra searches only if the first one doesn't turn up anything.

File: lib/common/results.c [hunk 2]
```diff
@@ -525,6 +526,7 @@ crm_exit_str(crm_exit_t exit_code)
         case CRM_EX_EXPIRED: return "Requested item has expired";
         case CRM_EX_NOT_YET_IN_EFFECT: return "Requested item is not yet in effect";
         case CRM_EX_INDETERMINATE: return "Could not determine status";
+        case CRM_EX_UNSATISFIED: return "Requested item does not satisfy constraints";
```
Comment (kgaillot) [inline lib/common/results.c]: To avoid confusion with user-configured constraints, I'd reword this, something like "Current conditions do not satisfy requirements" or "Not applicable under current conditions".

File: tools/crm_rule.c (hunk 1 above)
Comment (clumens) [reply tools/crm_rule.c]: Okay, I can do that.  Would you like more tests for all these possibilities, too?

File: lib/common/results.c (hunk 2 above)
Comment (clumens) [reply lib/common/results.c]: Good point.  Fixed.

**Comment by clumens** [comment]
> This would be a good step forward, but I think we also need pe_eval_date_expression() to return pe_date_{after,before}_range for date_spec with years.

Isn't that already happening at line 662, where pe_cron_range_satisfied gets called?  I also don't know how my tests would succeed/fail as appropriate if this weren't already happening.

File: tools/crm_rule.c (hunk 1 above)
Comment (kgaillot) [reply tools/crm_rule.c]: Might as well

**Comment by kgaillot** [comment]
> > This would be a good step forward, but I think we also need pe_eval_date_expression() to return pe_date_{after,before}_range for date_spec with years.
> 
> Isn't that already happening at line 662, where pe_cron_range_satisfied gets called? I also don't know how my tests would succeed/fail as appropriate if this weren't already happening.

Currently the test returns "Rule cli-prefer-rule-dummy-date_spec-only-years does not satisfy conditions" for the 2019 test, but it should return "expired". With years we can know whether it's before/after.

**Comment by clumens** [comment]
> Currently the test returns "Rule cli-prefer-rule-dummy-date_spec-only-years does not satisfy conditions" for the 2019 test, but it should return "expired". With years we can know whether it's before/after.

Just a quick clarification - should crm_rule now be able to check date_expressions that have year= plus other stuff (but not moon=), or just year=?

**Comment by kgaillot** [comment]
> Just a quick clarification - should crm_rule now be able to check date_expressions that have year= plus other stuff (but not moon=), or just year=?

Right, year + anything but moon.

**Comment by kgaillot** [comment]
It should be straightforward when everything is specified from least specific to most specific, e.g. years and months. Not so much if intermediates are unspecified, e.g. mondays in 2020. I guess to be consistent we can leave those as satisfied/unsatisfied only, since they could be partially in the past and partially in the future. Although if all occurrences are in the past, we could return expired.

**Comment by clumens** [comment]
I haven't nearly tested the new pe_cron_range_satisfied to my satisfaction, but here's a start.

File: tools/crm_rule.c [hunk 3]
```diff
@@ -63,24 +63,66 @@ crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effe
...
-    /* Get all rules matching the given ID, which also have only a single date_expression
-     * child whose operation is not 'date_spec'.  This is fairly limited, but it's hard
-     * to check expressions more complicated than that.
+    /* Get all rules matching the given ID which are also simple enough for us to check.
+     * For the moment, these rules must only have a single date_expression child and:
+     * - Do not have a date_spec operation, or
+     * - Have a date_spec operation that contains years= but does not contain moon=.
+     *
+     * We do this in steps to provide better error messages.  First, check that there's
+     * any rule with the given ID.
```
Comment (kgaillot) [inline tools/crm_rule.c]: Let's separate this into two commits, one for breaking the xpath into steps and the other for processing years

I only say this because lately I've been annoyed looking at some of my own past commits and thinking I really should have split them into three or four :)

File: tools/crm_rule.c [hunk 4]
```diff
@@ -63,24 +63,66 @@ crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effe
...
+    free(xpath);
+    freeXpathObject(xpathObj);
+
+    /* Next, make sure it has exactly one date_expression. */
+    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression", rule_id);
+    xpathObj = xpath_search(cib_constraints, xpath);
+    max = numXpathResults(xpathObj);
+
+    if (max != 1) {
+        CMD_ERR("Rule must have exactly one date_expression");
```
Comment (kgaillot) [inline tools/crm_rule.c]: Minor rephrase, something like "Can't check rule %s because it has more than one date_expression" -- I don't want to suggest that there's something wrong with the rule itself

File: tools/crm_rule.c [hunk 5]
```diff
@@ -63,24 +63,66 @@ crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effe
...
     max = numXpathResults(xpathObj);
+
     if (max == 0) {
-        CMD_ERR("No rule found with ID=%s containing a date_expression", rule_id);
+        CMD_ERR("No rule found with ID=%s", rule_id);
         rc = -ENXIO;
         goto bail;
     } else if (max > 1) {
-        CMD_ERR("More than one date_expression in %s is not supported", rule_id);
+        CMD_ERR("More than one rule with ID=%s found", rule_id);
```
Comment (kgaillot) [inline tools/crm_rule.c]: FYI schema validation should normally make this impossible, but this is worthwhile as a backstop, especially since a user could theoretically supply a manually edited CIB file that didn't go through validation

File: tools/crm_rule.c [hunk 6]
```diff
@@ -63,24 +63,66 @@ crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effe
...
+        free(xpath);
+        freeXpathObject(xpathObj);
+
+        xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation='date_spec' and date_spec/@years and not(date_spec/@moon)]",
+                                  rule_id);
+        xpathObj = xpath_search(cib_constraints, xpath);
+        max = numXpathResults(xpathObj);
+
+        if (max == 0) {
+            CMD_ERR("Rule either must not use date_spec, or use date_spec with year= but not moon=");
```
Comment (kgaillot) [inline tools/crm_rule.c]: years

File: lib/pengine/rules.c [hunk 7]
```diff
@@ -466,73 +466,110 @@ decodeNVpair(const char *srcstring, char separator, char **name, char **value)
...
+    free(value_high);
+
+    if (rc == pe_date_within_range) {
+        crm_debug("Condition '%s' in %s: passed", value, xml_field);
+    } else {
+        crm_debug("Condition '%s' in %s: failed", value, xml_field);
+    }
+
+    return rc;
+}
```
Comment (kgaillot) [inline lib/pengine/rules.c]: Less weird is better :)

File: tools/crm_rule.c (hunk 5 above)
Comment (clumens) [reply tools/crm_rule.c]: Yeah, agreed.  I realized this was likely an impossible situation when trying to write a test for this case.  Still, I like having it there.

**Comment by clumens** [comment]
I've updated this to address the review comments, but it's not ready to be merged yet.  I want #1987 to go in first so I can then rebase on top of that and verify that everything's still working.

**Comment by clumens** [comment]
Now that the unit tests are in, I've updated this PR.  I've done so in a new WIP patch at the end, because there's several return values of pe_cron_range_satisfied that I'm not really happy with.  Check them out and let me know what you think.

**Comment by clumens** [comment]
On second thought, maybe the return values are almost entirely correct.  pe_cron_range_satisfied appears to return `pe_date_op_satisfied` on success and anything other `pe_date_*` value on failure.  The more detailed failure codes let you know exactly what the problem was.  `yeardays_doesnt_satisfy` is still weird though.

**Comment by clumens** [comment]
Updated to fix a typo in an intermediate patch.  No one ever would have seen it in the finished product, but might as well get rid of it.  Also, I've decided that `yeardays_doesnt_satisfy` is correct too.

File: lib/pengine/tests/rules/pe_cron_range_satisfied.c [hunk 8]
```diff
@@ -16,99 +16,101 @@ run_one_test(const char *t, const char *x, gboolean expected) {
...
+    run_one_test("2000-01-01", "<date_spec id='spec' years='2010-2030'/>", pe_date_before_range);
 }
 
 static void
 time_after_year_range(void) {
-    run_one_test("2020-01-01", "<date_spec id='spec' years='2010-2015'/>", FALSE);
+    run_one_test("2020-01-01", "<date_spec id='spec' years='2010-2015'/>", pe_date_after_range);
 }
 
+/* IS THIS RIGHT? */
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pengine/tests/rules/pe_cron_range_satisfied.c]: I don't mind working on decodeNVpair (and I assume check_one as well).  Ideally, I would make it support open-ended ranges in both directions.  A range with no start would be equivalent to end=, and a range with no end would be equivalent to start=.  A single number without a dash would just be a single number.

I haven't looked too closely at it, but I'm wondering if at least one of those would be a break in behavior, given the comment in check_one if value_high_i is -1.

File: cts/cli/regression.rules.exp [hunk 9]
```diff
@@ -17,7 +17,12 @@ Error checking rule: No such device or address
       <primitive class="ocf" id="dummy" provider="heartbeat" type="Dummy"/>
     </resources>
     <constraints>
-      <rsc_location id="no-date-expression" rsc="dummy" score="-INFINITY" node="node01"/>
+      <rsc_location id="cli-too-many-date-expressions" rsc="dummy">
+        <rule id="cli-rule-too-many-date-expressions" score="INFINITY" boolean-op="or">
+          <date_expression id="cli-date-expression-1" operation="gt" start=""/>
+          <date_expression id="cli-date-expression-2" operation="lt" end=""/>
+        </rule>
+      </rsc_location>
```
[the remark this answers is not in the document]
Comment (clumens) [reply cts/cli/regression.rules.exp]: Sorry, I'm not quite following here.

File: lib/pengine/tests/rules/pe_cron_range_satisfied.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pengine/tests/rules/pe_cron_range_satisfied.c]: I worked on this a bit.  See #1991.  I can do more there if it'll be useful.  I'd like to get that one in, and then come back and revisit this PR.

File: cts/cli/regression.rules.exp (hunk 9 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply cts/cli/regression.rules.exp]: A test for a rule without any date_expression is still a good idea.

A &lt;rule&gt; can contain any combination of &lt;expression&gt; &lt;date_expression&gt; and (sub-) &lt;rule&gt; elements. &lt;expression&gt; checks node attributes:

https://clusterlabs.org/pacemaker/doc/en-US/Pacemaker/2.0/html-single/Pacemaker_Explained/index.html#idm47160816617536

I didn't think about sub-rules before, we should use "//" instead of "/" when xpath'ing for date_expression.

**Comment by clumens** [comment]
Rebased on top of #1991.  I still need to address https://github.com/ClusterLabs/pacemaker/pull/1982#discussion_r382658716.

File: lib/pengine/rules.c [hunk 10]
```diff
@@ -443,74 +443,106 @@ phase_of_the_moon(crm_time_t * now)
...
 
-gboolean
-pe_cron_range_satisfied(crm_time_t * now, xmlNode * cron_spec)
-{
-    const char *value = NULL;
+    if (pcmk__parse_ll_range(value, &low, &high) == pcmk_rc_unknown_format) {
+       goto bail;
+    } else if (low == -1 && high == -1) {
+        /* A single number was given, not a range. */
+       low = crm_parse_ll(value, NULL);
```
Comment (kgaillot) [inline lib/pengine/rules.c]: Does this need to be rebased on current master? We had reworked this section ...

File: lib/pengine/rules.c [hunk 11]
```diff
@@ -443,74 +443,106 @@ phase_of_the_moon(crm_time_t * now)
...
+           rc == pe_date_result_undetermined;
+}
 
-    crm_time_get_timeofday(now, &h, &m, &s);
+#define CHECK_ONE(spec, name, var) { \
+    pe_eval_date_result_t subpart_rc = check_one(spec, name, var); \
+    if (check_passes(subpart_rc) == FALSE) { \
+        return subpart_rc; \
+    } \
+}
```
Comment (kgaillot) [inline lib/pengine/rules.c]: Just as a general best practice, we wrap multi-line macros in do { ... } while (0). That construct can be used everywhere a single function call can, so it avoids non-obvious usage issues. E.g.

    if (a)
        CHECK_ONE(a, b, c);
    else
        return;

wouldn't compile because it would expand to if { ... }; else return; and the semi-colon would make the else hanging. Of course our style is to always use braces but it's still a good general practice.

File: tools/crm_rule.c [hunk 12]
```diff
@@ -135,22 +135,22 @@ crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effe
...
-        rc = 0;
+        rc = pcmk_rc_ok;
     } else if (result == pe_date_op_satisfied) {
         printf("Rule %s satisfies conditions\n", rule_id);
-        rc = 0;
+        rc = pcmk_rc_ok;
     } else if (result == pe_date_after_range) {
         printf("Rule %s is expired\n", rule_id);
-        rc = 1;
+        rc = -1;
```
Comment (kgaillot) [inline tools/crm_rule.c]: While this would work, it's treading on shaky ground. Negative values should be presumed reserved, whether for pacemaker (<= -1001) or for external callers. We could either add new pcmk_rc_ values, or have the function return crm_exit_t directly. Since pe_eval_date_result_t is internal, we could even potentially replace it with pcmk_rc_ entirely.

File: cts/cts-cli.in [hunk 13]
```diff
@@ -932,7 +932,19 @@ test_rules() {
...
+<rsc_location id="cli-too-many-date-expressions" rsc="dummy">
+  <rule id="cli-rule-too-many-date-expressions" score="INFINITY" boolean-op="or">
+    <date_expression id="cli-date-expression-1" operation="gt" start="2020-01-01 01:00:00 -0500"/>
+    <date_expression id="cli-date-expression-2" operation="lt" end="2019-01-01 01:00:00 -0500"/>
+  </rule>
+</rsc_location>
+EOF
+
+    cibadmin -C -o constraints -x "$TMPXML"
+    rm -f "$TMPXML"
```
Comment (kgaillot) [inline cts/cts-cli.in]: Definitely not needed for this PR, but just FYI, cibadmin also supports -p to get the XML from stdin so we could do echo '...' | cibadmin -p ...

File: lib/pengine/rules.c (hunk 10 above)
Comment (clumens) [reply lib/pengine/rules.c]: Yeah, good catch.  And there was another bug in this area too so it's a good thing I had to take a second look.

File: tools/crm_rule.c (hunk 12 above)
Comment (clumens) [reply tools/crm_rule.c]: Oh, I thought the purpose of the -1 to -1000 reserved space was that callers could use whatever values in that range as a kind of one-off set of return values.  I guess of those options, I'd prefer to merge pe_eval_date_result_t into pcmk_rc_e.  I'd like to not have functions return crm_exit_t if I can help it.

File: cts/cts-cli.in (hunk 13 above)
Comment (clumens) [reply cts/cts-cli.in]: Yeah, good to know.  I was just trying to fit in with everything else here.

File: include/crm/common/results.h [hunk 14]
```diff
@@ -107,6 +107,12 @@ enum pcmk_rc_e {
     /* When adding new values, use consecutively lower numbers, update the array
      * in lib/common/results.c, and test with crm_error.
      */
+    pcmk_rc_after_range         = -1026,
+    pcmk_rc_within_range        = -1025,
+    pcmk_rc_before_range        = -1024,
+    pcmk_rc_undetermined        = -1023,
+    pcmk_rc_op_satisfied        = -1022,
```
Comment (kgaillot) [inline include/crm/common/results.h]: I'm thinking "satisfied" should just use pcmk_rc_ok.

In fact looking at the code again, the only place we distinguish pcmk_rc_within_range from pcmk_rc_op_satisfied is in crm_rule_check(), to print "is still in effect" vs "satisfies conditions". I'm thinking that means the same thing anyway, so maybe we could use pcmk_rc_ok for both cases.

In any case, both cases need to map to CRM_EX_OK (using pcmk_rc_ok would get that automatically)

File: lib/common/results.c [hunk 15]
```diff
@@ -216,6 +216,30 @@ static struct pcmk__rc_info {
       "IPC server is blocked by unauthorized process",
       -pcmk_err_generic,
     },
+    { "pcmk_rc_op_unsatisifed",
+      "Operation unsatisfied",
```
Comment (kgaillot) [inline lib/common/results.c]: Might as well as use "Not applicable under current conditions" to be consistent with the exit code

File: include/crm/common/results.h (hunk 14 above)
Comment (clumens) [reply include/crm/common/results.h]: I guess I'd thought pcmk_rc_op_satisfied and pcmk_within_range might be more broadly useful later, in other areas of the code.  We might somewhere else want to know whether something's before/within/after a range without assigning a judgement of whether that's ok or not.  pcmk_rc_op_satisfied seems less useful, though.

Comment (clumens) [reply include/crm/common/results.h]: The latest push gets rid of pcmk_rc_op_satisfied, replacing it with pcmk_rc_ok.
